brainfuck/brainfuck.go

```diff
@@ -1,66 +1,74 @@
 package brainfuck
 
-import "fmt"
-
-var executableCommands = map[rune]func(*memmory){
-	'+': func(mem *memmory) {
-		mem.cells[mem.pointer]++
-	},
-	'-': func(mem *memmory) {
-		mem.cells[mem.pointer]--
-	},
-	'>': func(mem *memmory) {
-		mem.pointer++
-	},
-	'<': func(mem *memmory) {
-		mem.pointer--
-	},
-	'.': func(mem *memmory) {
-		fmt.Printf("%c", mem.cells[mem.pointer])
-	},
-	',': func(mem *memmory) {
-		fmt.Scanf("%c", &mem.cells[mem.pointer])
-	},
-	'[': func(mem *memmory) {
-		if mem.cells[mem.pointer] == 0 {
-			if loopsStack[0].openIndex == codePointer {
-				loopsStack[0].executing = false
-			}
-		} else {
-			if loopsStack[0].openIndex != codePointer {
-				loopsStack = append(make([]loop, 1), loopsStack...)
-				loopsStack[0].openIndex = codePointer
-				loopsStack[0].executing = true
-			}
-		}
-	},
-	']': func(mem *memmory) {
-		if mem.cells[mem.pointer] == 0 || !loopsStack[0].executing {
-			loopsStack = loopsStack[1:]
-		} else {
-			// - 1 here is necessary because we increment codePointer each iteration
-			codePointer = loopsStack[0].openIndex - 1
-		}
-	},
+// This map contains empty structures that inherit command interface
+var executableCommands = map[rune]command{
+	// increment
+	'+': increment{},
+	// decrement
+	'-': decrement{},
+	// increment data pointer
+	'>': incrementDataPointer{},
+	// decrement data pointer
+	'<': decrementDataPointer{},
+	// output
+	'.': output{},
+	// input
+	',': input{},
+	// the beginning of loop
+	'[': loop{},
+	// the end of loop
+	']': endLoop{},
+	// All the functions after the ']', are not implemented in original Brainfuck language
+	// copy
+	'c': copy{},
+	// paste
+	'p': paste{},
+	// clear
+	'0': zero{},
 }
 
+// The memmory set we're going to use everywhere.
 var memmorySet memmory
-var codePointer int
-var loopsStack []loop
 
-type loop struct {
-	openIndex int
-	executing bool
+// This slice of loops we're going to use as a stack for
+// loop commands in which we will append commands inside "[]".
+//
+// After the interpretation we will execute it, but not with the
+// currentLoop[0].execute(memmorySet), because we don't need to repeat it
+var currentLoop = []loop{
```

Comment:
why are you using global var? I think it's not suitable in this case. It's better to return `currentLoop` and then pass it to `compile(currentLoop)`